### PR TITLE
Hand in

### PR DESCRIPTION
**Pull request by Guy-Markman**
@alonbl
Is this how you want it?

### CODE REVIEW

File: EX04/02-sigchld.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+def child_process():
+    start = time.time()
+    time.sleep(TIME_SLEEP_CHILD)
+    print time.time() - start
+
+
+def parent_process(child_pid):
+    util.write_to_target(
+        1, "Parent pid: %s\nChild pid: %s\n" % (os.getpid(), child_pid)
```
Comment (alonbl) [inline EX04/02-sigchld.py]: ---

File: EX04/02-sigchld.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+        )
+    time.sleep(TIME_SLEEP_PARENT)
```
Comment (alonbl) [inline EX04/02-sigchld.py]: is this sleeping 60 seconds?

File: EX04/03-sigalrm.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+def child_process():
+    time.sleep(TIME_SLEEP_CHILD)
+    sys.exit()
+
+
+def parent_process(child_pid):
+    signal.signal(signal.SIGCHLD, handler_SIGCHLD)
+    util.write_to_target(
+        1, "Parent pid: %s\nChild pid: %s\n" % (os.getpid(), child_pid)
```
Comment (alonbl) [inline EX04/03-sigalrm.py]: ...

File: EX04/03-sigalrm.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    time.sleep(TIME_SLEEP_PARENT)
+    sys.exit()
+
+
+def handler_SIGCHLD(signum, frame):
+    os.wait()
+
+
+def handler_SIGALRM(signum, frame):
+    print "waiting"
```
Comment (alonbl) [inline EX04/03-sigalrm.py]: how many types do you see this?

File: EX04/04-eintr.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+import signal
+
+TIME_SLEEP_CHILD = 2
+TIME_SLEEP_PARENT = 60
+
+BUFF_SIZE = 1000
+
+
+def child_process():
+    util.read_from_target(0, BUFF_SIZE)
```
Comment (alonbl) [inline EX04/04-eintr.py]: should be done in parent

and where do you address the ENTR?

File: EX04/05-ctrl-c.py [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+
+
+def hal9000(signum, frame):
+    util.write_to_target(1, "I'm sorry Dave, I'm afraid I can't do that\n")
+
+
+def main():
+    signal.signal(signal.SIGINT, hal9000)
+    while True:
+        continue
```
Comment (alonbl) [inline EX04/05-ctrl-c.py]: you are wasting CPU!!!

File: EX04/08-daemon.py [hunk 7]
```diff
@@ -0,0 +1,10 @@
+#!/usr/bin/env python
+import signals
+import os
+
+
+def main():
```
Comment (alonbl) [inline EX04/08-daemon.py]: ?

File: EX04/04-eintr.py (hunk 5 above)
Comment (Guy-Markman) [inline EX04/04-eintr.py]: What do you mean by "address the ENTR"?

**Comment by alonbl** [comment]
> What do you mean by "address the ENTR"?

Question was how to deal with EINTER state, no? where do you handle this
state?

**Comment by Guy-Markman** [comment]
Please take a look at util.read_from_target. I deal with it over there

**Comment by alonbl** [comment]
On 9 October 2016 at 16:59, Guy Markman <notifications@github.com> wrote:
>
> Please take a look at util.read_from_target. I deal with it over there
>

Oh... now I see, good.

**Comment by Guy-Markman** [comment]
As agrred with @alonbl 
I'm merging and will add daemon soon

**Pull request merged**